Scheduler trait + transition all boards to round robin scheduler

hudson-ayers opened the pull request:
### Pull Request Overview

This pull request creates a Scheduler trait that allows for different scheduling algorithms to be used by the Kernel, and implements 4 schedulers using this new trait:

1.  A round robin scheduler which resolves a number of the problems with the current scheduling implementation, such that it is no longer possible for any app to starve another of the CPU indefinitely, and which fairly distributes timeslices even when interrupts arrive during execution.
2. A cooperative scheduler, which can be used on platforms without a systick.
3. A multi-level feedback queue scheduler which defers bottom-half interrupt handling until timeslice expiration or yield. This shows more complex scheduling algorithms are possible, that storing per-process state is possible, and that this design allows for schedulers to choose when to process interrupts.

This PR puts each scheduling algorithm into a `sched/` folder in the Kernel crate, and creates components for each scheduler.

I am interested in this from the perspective of wanting to eventually implement a real-time scheduler of some sort, which may require the scheduler to have fine-tuned control of when to handle interrupts. I could also imagine a power-aware scheduler, which would need the ability to control when to call chip.sleep(), among other things. Implementing a priority scheduler may require the ability to run different apps with different time quantums. 

### Testing Strategy

Tested several combinations of apps, including whileone, on each scheduler. 


### TODO or Help Wanted

We need to decide which schedulers are worth keeping upstream, as code rot is inevitable for schedulers not used.


### Documentation Updated

- [ ] Once this interface is finalized, will need to add documentation describing how/why to choose different schedulers.

### Formatting

- [x] Ran `make formatall`.

hudson-ayers commented:
While I haven't implemented this for all the boards yet, it really isn't a WIP -- I am just waiting for feedback on the design. If people like this, I transition all the boards to this design and finalize this PR.

ppannuto commented on kernel/src/sched/mod.rs:
```diff
@@ -0,0 +1,474 @@
+
+    /// Helper function for determining if we should service processes or go to
+    /// sleep.
+    fn processes_blocked(&self) -> bool {
```
I'm not sure if this is just a copy/paste or a new interface, but this is a confusing name. Something like `any_outstanding_work` I think would be clearer.

hudson-ayers replied:
this is from the old interface, and I made changes so that my PR now doesn't touch this function.

alistair23 commented on arch/cortex-m/src/systick.rs:
```diff
@@ -142,6 +149,12 @@ impl kernel::SysTick for SysTick {
         SYSTICK_BASE.syst_cvr.set(0);
     }
 
+    fn pause(&self) {
```
I don't think RISC-V can pause it's counter, so I'm not sure how this would work in RV32.

hudson-ayers replied:
looks like you are right, I will need to rework this :/ it was so elegant for arm..

alistair23 commented on boards/components/src/round_robin.rs:
```diff
@@ -0,0 +1,73 @@
+//! Component for a round robin scheduler.
```
Should all of the scheduler components maybe go in another `scheduler` subdir/sub-crate?

alistair23 commented on kernel/src/sched/cooperative.rs:
```diff
@@ -0,0 +1,170 @@
+                    process.setup_mpu();
+                    chip.mpu().enable_mpu();
+                    let context_switch_reason = process.switch_to();
+                    chip.mpu().disable_mpu();
```
This will need to be rebased on: https://github.com/tock/tock/pull/1822

hudson-ayers replied:
I think that #1822 changed such that I no longer have to rebase after all

alistair23 commented on kernel/src/sched/mlfq.rs:
```diff
@@ -0,0 +1,309 @@
+    const MIN_QUANTA_THRESHOLD_US: u32 = 500;
+    /// How often to restore all processes to max priority
+    pub const PRIORITY_REFRESH_PERIOD_MS: u32 = 5000;
+    pub const NUM_QUEUES: usize = 3;
```
Should these be settable from the board that creates it?

hudson-ayers commented on boards/arty-e21/src/main.rs:
```diff
@@ -2,15 +2,15 @@
 #![no_std]
 #![no_main]
-#![feature(const_fn, in_band_lifetimes)]
+#![feature(const_fn, in_band_lifetimes, const_in_array_repeat_expressions)]
```
Without this feature I was unable to write components for scheduler initialization that are generic over the number of processes a board supports. If a board did not want to use this feature, it could initialize the scheduler without using the component.

ppannuto replied:
```suggestion
#![feature(const_fn, in_band_lifetimes)]
// Required to support components that are generic over the number of processes
// a board supports. To avoid this feature, boards can initialize the scheduler directly
// without using the component interface.
//
// Tock expects this feature to stabilize in the medium future.
// Tracking: https://github.com/rust-lang/rust/issues/49147
#![feature(const_in_array_repeat_expressions)]
```

That makes sense. Maybe we should start a "documenting why / path to stable" on our features? I know this isn't the ideal place to prototype this out, but it's where it came to mind.

hudson-ayers replied:
this seems reasonable, but would we have to put that disclaimer in every main.rs?

ppannuto replied:
mhm, maybe we should centralize them somewhere instead, maybe a in the stabilization tracking issue

hudson-ayers commented on kernel/src/sched/round_robin.rs:
```diff
@@ -0,0 +1,203 @@
+            systick.reset();
+            systick.set_timer(Self::DEFAULT_TIMESLICE_US);
+        }
+        systick.enable(false); //resumes counting down
```
The current scheduler takes the same approach of charging kernel overhead to process time quanta. In the case that a process becomes ready because of a callback, but is not yet in the Running state, this means that overhead of handling that callback is charged to the process before the process actually begins executing.

I could go either way on which approach is better.

ppannuto replied:
I'm not enough of a schedulers / real-time person to know offhand, but I imagine this decision is one that's well defined in the theory. Thinking a little more, my instinct is that to ensure schedule reasoning, this time has to be accounted for somewhere, but I could believe that the standard is to charge it to the "kernel thread" or something (or certainly it could be to the process).

Ping @phil-levis, is this something that's standardized that you know of?

jrvanwhy replied:
I do not have nearly the historical knowledge that Phil has, but I do know quite a bit about realtime systems, so I'll but in anyway :-)

I can only think of two situations where a thread would hit its timeslice limit in a well-designed realtime system:

1. The thread is a background thread that performs non-realtime work, such as logging. This case can be treated the same as non-realtime systems.

1. The thread is a realtime thread that overran its deadline (i.e. it is buggy). In this case, the timeslice exists to protect *other* realtime threads from the buggy realtime thread. The other threads care about total latency, which is the sum of the kernel and userspace time that preempted them, so both should count towards the timeslice limit.

hudson-ayers replied:
From some reading today, as far as I can tell the Linux scheduler charges in-kernel work (such as handling syscalls) to the userspace thread that made the syscall.

hudson-ayers commented on kernel/src/sched/round_robin.rs:
```diff
@@ -0,0 +1,203 @@
+//! Round Robin Scheduler for Tock
+//! This scheduler is specifically a Round Robin Scheduler with Interrupts.
```
All the pointers I can find are to lecture slides like this one: https://www.eecs.umich.edu/courses/eecs461/lecture/SWArchitecture.pdf is that an acceptable pointer or would it need to be a textbook

ppannuto replied:
🤷 , that text is probably fine. The goal is just to map that label onto something that people who aren't familiar with scheduler theory would be able to quickly understand what the goals of the scheduler are.

bradjc commented:
Having S copies of `do_process()` worries me. There are very subtle bugs that come from that function, and the way that github PRs are displayed, it is very easy to forget to update the copies since Github won't show that the other code _hasn't_ changed. There is already a regression (we have to check for pending interrupts _and_ deferred calls in do_process()).

I also want to know what you think about leaving sched.rs in this PR so that the diff is more clear. More re-org can happen in a separate PR that only moves files and doesn't also introduce new code.

hudson-ayers commented:
> I also want to know what you think about leaving sched.rs in this PR so that the diff is more clear. More re-org can happen in a separate PR that only moves files and doesn't also introduce new code.

at around the same time you made this comment I force pushed a change that moved sched/mod.rs back to sched.rs because rust treats those namings as equivalent, does that work?

hudson-ayers commented:
> Having S copies of do_process() worries me. There are very subtle bugs that come from that function, and the way that github PRs are displayed, it is very easy to forget to update the copies since Github won't show that the other code hasn't changed. There is already a regression (we have to check for pending interrupts and deferred calls in do_process()).

Yeah, I think this is an important concern. The challenge is that different schedulers might want to do different things in `do_process`, (cooperative scheduler does not need to do anything with systick, round_robin needs to track whether process is being rescheduled, mlfq defers bottom half interrupt handling until systick expiration or yield) and I was not sure how to encapsulate those different options in a no-cost manner. Especially because future schedulers might have additional stuff they want to do in do_process that we have not yet imagined.

One option might be to decide we are going to constrain the interface for now and say that you can't defer interrupt handling and the only way to go without a systick is by passing the dummy systick at initialization, and then leave `do_process()` in sched.rs

hudson-ayers commented:
I spent the last few days reworking this interface, now there are not S copies of `do_process()`. I did make modifications to `do_process()` to make it a little more flexible, at some (hopefully low) cost. I think the resulting design is a lot cleaner, though probably slightly less efficient in some cases.

hudson-ayers commented:
I have now applied the necessary changes to all boards so this passes CI.

I believe I have addressed most of the previously mentioned concerns, aside from the introduction of the unstable feature constants_in_array_repeat_expressions, which 

  - can be removed by not using the component and initializing manually
  - could be replaced by a hacky recursive macro (I would rather keep this idiomatic in the likelihood this feature becomes stable before we remove all others)

Notably, I moved all risc-v boards over to use the cooperative scheduler, as a non cooperative scheduler does not make sense for boards that do not have a systick implemented.

This still needs some pretty rigorous testing, but is ready for review.

ppannuto commented on arch/cortex-m/src/systick.rs:
```diff
@@ -132,6 +132,13 @@ impl kernel::SysTick for SysTick {
 
+    fn get_value(&self) -> u32 {
+        let tics = SYSTICK_BASE.syst_cvr.read(CurrentValue::CURRENT);
+        //let hertz = self.hertz();
+        //tics * 1_000_000 / hertz
```
```suggestion
```

?

hudson-ayers replied:
woah, those were supposed to be included. I took them out when benchmarking this function and guess I forgot to bring them back. good catch.

ppannuto commented on boards/components/src/sched/cooperative.rs:
````diff
@@ -0,0 +1,72 @@
+//!
+//! Usage
+//! -----
+//! (Must be final lines in main.rs)
+//! ```rust
+//! let scheduler = components::cooperative::CooperativeComponent::new(board_kernel, &PROCESSES)
+//!     .finalize(components::coop_component_helper!(NUM_PROCS));
+//! scheduler.kernel_loop(&imix, chip, Some(&imix.ipc), &main_cap);
````
```suggestion
//! Note that `kernel_loop` never returns, thus this must end main.rs
//! ```rust
//! let scheduler = components::cooperative::CooperativeComponent::new(board_kernel, &PROCESSES)
//!     .finalize(components::coop_component_helper!(NUM_PROCS));
//! scheduler.kernel_loop(&imix, chip, Some(&imix.ipc), &main_cap)
```

ppannuto commented on boards/components/src/sched/mlfq.rs:
````diff
@@ -0,0 +1,99 @@
+//!
+//! Usage
+//! -----
+//! (Must be final lines in main.rs)
+//! ```rust
+//! let scheduler =
+//!     components::mlfq::MLFQComponent::new(board_kernel, mux_alarm, &PROCESSES).finalize(
+//!         components::mlfq_component_helper!(sam4l::ast::Ast, NUM_PROCS),
+//!     );
+//! scheduler.kernel_loop(&imix, chip, Some(&imix.ipc), &main_cap);
````
```suggestion
//! Note that `kernel_loop` never returns, thus this must end main.rs
//! ```rust
//! let scheduler =
//!     components::mlfq::MLFQComponent::new(board_kernel, mux_alarm, &PROCESSES).finalize(
//!         components::mlfq_component_helper!(sam4l::ast::Ast, NUM_PROCS),
//!     );
//! scheduler.kernel_loop(&imix, chip, Some(&imix.ipc), &main_cap)
```

ppannuto commented on boards/components/src/sched/mlfq.rs:
```diff
@@ -0,0 +1,99 @@
+//! Component for a multi-level feedback queue scheduler.
+//!
+//! This provides one Component, MLFQComponent.
```
```suggestion
//! This provides one Component, MLFQComponent.
//! https://en.wikipedia.org/wiki/Multilevel_feedback_queue
```

ppannuto commented on boards/components/src/sched/priority.rs:
````diff
@@ -0,0 +1,36 @@
+//!
+//! Usage
+//! -----
+//! (Must be final lines in main.rs)
+//! ```rust
+//! let scheduler =
+//!     components::priority::PriorityComponent::new(board_kernel).finalize(());
+//! scheduler.kernel_loop(&imix, chip, Some(&imix.ipc), &main_cap);
````
```suggestion
//! Note that `kernel_loop` never returns, thus this must end main.rs
//! ```rust
//! let scheduler =
//!     components::priority::PriorityComponent::new(board_kernel).finalize(());
//! scheduler.kernel_loop(&imix, chip, Some(&imix.ipc), &main_cap)
```

ppannuto commented on boards/components/src/sched/round_robin.rs:
````diff
@@ -0,0 +1,74 @@
+//!
+//! Usage
+//! -----
+//! (Must be final lines in main.rs)
+//! ```rust
+//! let scheduler = components::round_robin::RoundRobinComponent::new(board_kernel, &PROCESSES)
+//!     .finalize(components::rr_component_helper!(NUM_PROCS));
+//! scheduler.kernel_loop(&imix, chip, Some(&imix.ipc), &main_cap);
````
```suggestion
//! Note that `kernel_loop` never returns, thus this must end main.rs
//! ```rust
//! let scheduler = components::round_robin::RoundRobinComponent::new(board_kernel, &PROCESSES)
//!     .finalize(components::rr_component_helper!(NUM_PROCS));
//! scheduler.kernel_loop(&imix, chip, Some(&imix.ipc), &main_cap)
```

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -284,63 +309,33 @@ impl Kernel {
         ipc: Option<&crate::ipc::IPC>,
-    ) {
-        let systick = chip.systick();
+        timeslice_us: u32,
```
```suggestion
        timeslice_us: Option<u32>,
```

Instead of having the cooperative (and other non-preempting schedulers) use the sentinel `0`, they should probably set `None` for this as they don't have "timeslices".

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -284,63 +309,33 @@ impl Kernel {
+    ) -> (StoppedExecutingReason, u32) {
         systick.reset();
-        systick.set_timer(KERNEL_TICK_DURATION_US);
+        systick.set_timer(timeslice_us);
```
Comment really for the next line.

I know this is "how it was", but I wonder if we need to enforce some minimum `timeslice_us` value. My concern is the universe where this value is small, and the timeslice manages to expire before the switch to userland completes. As I see it, we have two opportunities to miss the systick interrupt:
 (1) It expires before we call `enable(true)` below, so the timer expires without triggering the interrupt
 (2) It expires *during* the switch to user (i.e. after the `enable(true)` call but before the context switch actually occurs); the interrupt handler will mark the systick expired, but then we're back to the context switch code effectively without a systick set.

ppannuto replied:
Thinking about the later `systick.enable(true)` more, I think we need to clarify the systick HIL a bit:

```
    /// Enables the timer
    ///
    /// Enabling the timer will begin a count down from the value set with
    /// `set_timer`.
    ///
    ///   * `with_interrupt` - if set, an expiring timer will fire an interrupt.
    fn enable(&self, with_interrupt: bool);
```

It's unclear to me from that language if the second call to `enable` will *reset* the systick timer (I think it will?) or simply enable interrupts.

Since the process can stay in the loop here depending on the syscall, it could effectively grant itself infinite time.

I think my instinct is that there should be exactly one call to `systick.enable` at the beginning of `do_process`

hudson-ayers replied:
> Comment really for the next line.
> 
> I know this is "how it was", but I wonder if we need to enforce some minimum `timeslice_us` value. My concern is the universe where this value is small, and the timeslice manages to expire before the switch to userland completes. As I see it, we have two opportunities to miss the systick interrupt:
> (1) It expires before we call `enable(true)` below, so the timer expires without triggering the interrupt
> (2) It expires _during_ the switch to user (i.e. after the `enable(true)` call but before the context switch actually occurs); the interrupt handler will mark the systick expired, but then we're back to the context switch code effectively without a systick set.

I think that the check on `systick.greater_than(MIN_QUANTA_THRESHOLD_US)` effectively enforces a minimum timeslice of `MIN_QUANTA_THRESHOLD_US`

hudson-ayers replied:
> Thinking about the later `systick.enable(true)` more, I think we need to clarify the systick HIL a bit:
> 
> ```
>     /// Enables the timer
>     ///
>     /// Enabling the timer will begin a count down from the value set with
>     /// `set_timer`.
>     ///
>     ///   * `with_interrupt` - if set, an expiring timer will fire an interrupt.
>     fn enable(&self, with_interrupt: bool);
> ```
> 
> It's unclear to me from that language if the second call to `enable` will _reset_ the systick timer (I think it will?) or simply enable interrupts.
> 
> Since the process can stay in the loop here depending on the syscall, it could effectively grant itself infinite time.
> 
> I think my instinct is that there should be exactly one call to `systick.enable` at the beginning of `do_process`

The cortex-m implementation does not reset the timer, from looking at the register writes. I think that is the correct behavior, I will update the documentation to reflect that.

I do not think we want the possibility of the systick firing while we are in the kernel -- I do not totally understand the systick_handler assembly, but the documentation seems to indicate it redirects control flow to the switch to user function, and I do not think we want that to be possible while we are in the kernel. I think calling enable twice seems to make sense, though perhaps the systick trait could be changed to have seperate start_counter() and enable_interrupts() functions.

rajivr replied:
> I do not think we want the possibility of the systick firing while we are in the kernel

This possibility does exist. I was doing something else with our kernel, when I discovered it few days back. [Here](https://github.com/tock/tock/blob/313a5e12298b83b49cb95dd5d39baa5332bb1a13/arch/cortex-m/src/syscall.rs#L164) we might come out of the kernel because of an interrupt or a syscall, while the systick timer is still running. So, we'll capture `systick_expired` to be `0`. Then by the time we get to stopping systick timer, the systick interrupt could have fired, and `Systick_CurrentValue:CURRENT[23:0]` would have `Systick_Reload:RELOAD[23:0]`. So, the remaining time slice that gets captured *could* be wrong. Capturing remaining time-slice accurately is important for our use-case.

While I've not coded the implementation yet, to me it seems like the correct approach should be that we stop the systick timer in assembly within `svc_handler`, `generic_isr` and `systick_handler`, so we can accurately capture `Systick_CurrentValue:CURRENT[23:0]` value.

ppannuto replied:
Oy, I had not yet even considered the race between systick and other interrupt handlers. Here lies very subtle and rare bugs :(.

On ARM at least, I think it should be viable to configure the systick to be lower priority than other interrupts, so we can be sure that a systick interrupt won't fire while other handlers are running. I don't know enough about Risc-V to say for certain whether the same fix is viable there.

More generally, I'm concerned about how we express this pretty low-level requirement in a robust, cross-platform way such that all chips / arches get it right.

hudson-ayers replied:
> > I do not think we want the possibility of the systick firing while we are in the kernel
> 
> 
> 
> This possibility does exist. I was doing something else with our kernel, when I discovered it few days back. [Here](https://github.com/tock/tock/blob/313a5e12298b83b49cb95dd5d39baa5332bb1a13/arch/cortex-m/src/syscall.rs#L164) we might come out of the user-space because of an interrupt or a syscall, while the systick timer is still running. So, we'll capture `systick_expired` to be `0`. Then by the time we get to stopping systick timer, the systick interrupt could have fired, and `Systick_CurrentValue:CURRENT[23:0]` would have `Systick_Reload:RELOAD[23:0]`. So, the remaining time slice that gets captured *could* be wrong. Capturing remaining time-slice accurately is important for our use-case.
> 
> 
> 
> While I've not coded the implementation yet, to me it seems like the correct approach should be that we stop the systick timer in assembly within `svc_handler`, `generic_isr` and `systick_handler`, so we can accurately capture `Systick_CurrentValue:CURRENT[23:0]` value.
> 
> 

Hey @rajivr , in order to accurately capture the amount of time used, it sounds like you must already be using a modified scheduler? If so, is this interface sufficient to capture the behavior you want from your scheduler? 

I agree that currently it is possible for the reload value to be read by systick.get_value() in the case of the race you describe —fortunately the cortex-m implementation sets the reload value to be 0 until set_value() is called, so I think it captures the equivalent of if the systick interrupt had occurred before the return from userspace. Does this sound right?

That being said, my opinion is that it makes sense to charge syscall handling that occurs in the kernel to the process. If we stop the systick in svc_handler, I am not sure we can do so. Do you have a different opinion on this? Also there is the problem that I believe the risc-v systick does not support a mechanism for pausing the timer.

rajivr replied:
> Hey @rajivr , in order to accurately capture the amount of time used, it sounds like you must already be using a modified scheduler? If so, is this interface sufficient to capture the behavior you want from your scheduler?

Sorry for the delay in replying.

Our scheduler is still under development. We are taking a different approach with our scheduler. Our scheduler interface is largely inspired from,

https://github.com/xen-project/xen/blob/stable-4.12/xen/include/xen/sched-if.h#L159-L169

You can replace `vcpu` with `process`, and it provides a rather minimal scheduler interface to the rest of the kernel.

`do_schedule` is where most of the scheduler policy decisions can be taken. In our use-case, we'll only be implementing one scheduler, but Xen has multiple scheduler implementations. In case you are curious, check out the files `sched_{credit, credit2, null, rt}.c` in the tree.

> That being said, my opinion is that it makes sense to charge syscall handling that occurs in the kernel to the process. If we stop the systick in svc_handler, I am not sure we can do so. Do you have a different opinion on this? Also there is the problem that I believe the risc-v systick does not support a mechanism for pausing the timer.

While RISC-V is not our immediate roadmap, I did check out the documentation for `mtime` and `mtimecmp`, when I was thinking about the design. I do not have hands on experience with RISC-V, so please the next comment with a grain of salt. 

I don't think we necessarily have to stop the timer in the trap handler. In case of RISC-V the approach that could potentially work would be to disable the `MTIE` bit in the trap handler, and then copy out the `mtime` value into a known memory location.

So, if we returned back to the kernel due to a system call or an interrupt, we can then check this known memory location and compute how much time was spent in the user-mode.

hudson-ayers replied:
One issue is that 2 of our 3 currently supported RISC-V chips only have a single hardware timer (the mtimer), so for these chips we will likely need to implement the systick as an alarm on top of an AlarmMux that shares a hardware timer with other alarms in the kernel. This makes disabling MTIE while in the kernel unpalatable, because it is also required for interrupts unrelated to timeslice expirations

rajivr replied:
> One issue is that 2 of our 3 currently supported RISC-V chips only have a single hardware timer (the mtimer), so for these chips we will likely need to implement the systick as an alarm on top of an AlarmMux that shares a hardware timer with other alarms in the kernel. This makes disabling MTIE while in the kernel unpalatable, because it is also required for interrupts unrelated to timeslice expirations

OpenTitan [Earlgrey](https://docs.opentitan.org/hw/top_earlgrey/doc/) does seem to have only one timer. I think having just one timer peripheral is very limiting. I would still prefer having a dedicated timer for the scheduler. Virtual timer would add additional layer of complexity into an important code-path in the kernel.

hudson-ayers replied:
> OpenTitan [Earlgrey](https://docs.opentitan.org/hw/top_earlgrey/doc/) does seem to have only one timer. I think having just one timer peripheral is very limiting. I would still prefer having a dedicated timer for the scheduler. Virtual timer would add additional layer of complexity into an important code-path in the kernel.

The sifive e21 has the same limitation. I am hoping I can come up with an abstraction where most boards use a dedicated timer, but boards without one can implement their own lightweight virtualization and pass a SysTick to the Kernel that behaves like a dedicated timer.

hudson-ayers replied:
Marking this conversation as resolved. `MIN_QUANTA_THRESHOLD_US` presents as an effective minimum timeslice, and the possible race condition for detecting timeslice usage should be resolved by #1985 .

hudson-ayers commented on kernel/src/sched.rs:
```diff
@@ -600,5 +597,7 @@ impl Kernel {
             }
         }
         systick.reset();
+        let time_executed_us = u32::min(timeslice_us - systick.get_value(), timeslice_us); //min is to protect from wrapping systick value
```
good catch, it should be before the reset

bradjc commented:
The model that the `Kernel` object holds key system state and is given references to objects it needs to do core functions still resonates with me. To that end, I still think the trait should be something like:

```rust
pub trait Scheduler {
    /// Return the next process to run and the timeslice length for that process.
    /// A timeslice of 0 means do not use the systick.
    fn next(&self) -> (Option<&'static dyn process::ProcessType>, u32);

    /// Informs the scheduler of why the last process stopped executing.
    fn result(&self, result: StoppedExecutingReason, execution_time_us: u32);
}
```

Kernel loop looks something like:
```rust
pub fn kernel_loop<P: Platform, C: Chip, SC: Scheduler>(
        &self,
        platform: &P,
        chip: &C,
        ipc: Option<&ipc::IPC>,
        scheduler: &SC,
        _capability: &dyn capabilities::MainLoopCapability,
    ) -> ! {
        loop {
            unsafe {
                chip.service_pending_interrupts();
                DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());

                loop {
                    if chip.has_pending_interrupts()
                        || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
                        || self.processes_blocked()
                    {
                        break;
                    }
                    let (process, timeslice_us) = scheduler.next();
                    process.map(|process| {
                        let reason =
                            self.do_process(platform, chip, process, ipc, timeslice_us, true);
                        scheduler.result(reason.0, reason.1);
                    });
                }

                chip.atomic(|| {
                    if !chip.has_pending_interrupts()
                        && !DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
                        && self.processes_blocked()
                    {
                        chip.sleep();
                    }
                });
            };
        }
    }
```

My thinking:
1. The duplicated kernel_loop code is no longer duplicated.
2. Implementing a scheduler is a lot more straightforward. Rather than being responsible for the entire loop, there is much less a scheduler author needs to worry about.
3. The issues raised in the original post I think can be handled. Having the scheduler specify a timeslice handles the cooperative case and otherwise provides flexibility. For a power aware scheduler, it could return `None` when `next()` is called if the system should sleep. 
4. There is nothing enforcing that `kernel.kernel_loop()` _must_ be called. A scheduler can provide its own loop if it absolutely needed it. But rather than requiring that complexity in case the need might arise, we can make the cases we have simpler while not excluding the custom option in the future.

I did a very rough prototype of this in the sched-bradjc branch...just enough to make sure I could get this to compile. The `next()` function also needs to return if the kernel events should interrupt the process, or a better way to handle this is needed.

I also think it is more intuitive for do_process to use timeslice as a flag: 0 means don't use the systick. Having the scheduler pass in the systick timer is odd to me: why would a scheduler have multiple references to a timer? (I know the purpose was to be able to pass in a dummy systick, but from only reading the API I found it unclear.)

bradjc commented on kernel/src/sched.rs:
```diff
@@ -599,6 +619,11 @@ impl Kernel {
                 }
             }
         }
+        let time_executed_us = u32::min(
+            timeslice_us.unwrap_or(0) - systick.get_value(),
+            timeslice_us.unwrap_or(0),
+        ); //min is to protect from wrapping systick value
```
Doesn't this calculate the time _not_ used? Or does systick count down? Or am I just not reading this right?

hudson-ayers replied:
The ARM systick counts down (#1985 renames `get_value()` to `get_time_remaining_us()`, which makes it clearer)

bradjc commented on boards/hail/src/main.rs:
```diff
@@ -466,5 +467,8 @@ pub unsafe fn reset_handler() {
+    let scheduler =
+        components::sched::round_robin::RoundRobinComponent::new(board_kernel, &PROCESSES)
+            .finalize(components::rr_component_helper!(NUM_PROCS));
+    scheduler.kernel_loop(&hail, chip, Some(&hail.ipc), &main_loop_capability);
```
I'm not sure about changing all of the schedulers. At the very least the title of this PR should change. But I'm a little wary of changing the scheduler out from under the board authors.

hudson-ayers replied:
I don't like the idea of leaving a bunch of boards sticking around on something that is the equivalent of the current scheduler, mostly because the current scheduler allows for apps to indefinitely starve other apps, which is something that we sort of claim to have designed Tock to prevent (e.g. the evil while(1) example).

If people feel strongly about this I can change it back, but I really believe that universal use of a round robin scheduler would a preferable default.

I changed the title for now.

ppannuto replied:
I think I'm okay with changing schedulers as I don't see it as a change in _intended_ functionality. I thought the "default scheduler" of Tock was always intended to be round robin with pre-emptions. I see this more conceptually as a bugfix than as change if anything.

bradjc replied:
The current scheduler is still implemented after this PR, right? If we don't like it, then we shouldn't keep it.

I'm fine with just updating the PR title, but in general I do want to make sure we are being deliberate when changing settings that are delegated to main.rs.

hudson-ayers replied:
The priority scheduler in this PR no longer enforces timeslices -- it is a strict priority scheduler where the highest priority process will always be running (with the kernel thread -- aka bottom half interrupts and deferred calls -- being higher priority than all userspace processes). I think this is worth keeping because priority scheduling can be useful in some systems, and this scheduler does not pretend that all apps will be given a chance to run when that is not in fact the case.

bradjc commented on kernel/src/lib.rs:
```diff
@@ -45,7 +45,11 @@ pub use crate::platform::systick::SysTick;
+pub use crate::sched::mlfq::{MLFQProcessNode, MLFQSched};
+pub use crate::sched::priority::PrioritySched;
+pub use crate::sched::round_robin::{RoundRobinProcessNode, RoundRobinSched};
+pub use crate::sched::{Kernel, Scheduler};
```
If we didn't have the legacy sched.rs file, I don't think we would go with that name now. I think we should start improving this by creating a `schedulers` (or `scheduler`) directory, and putting the schedulers in there.

hudson-ayers replied:
Using `sched/` makes it so that the individual schedulers are in the same module as `sched.rs`, so they can use private functions defined in that file (e.g. `processes_blocked()`). If we want to move them into a `schedulers/` directory we either have to:

a) change `processes_blocked()`, `do_process()`, and `Kernel.processes` to `pub(crate)` visibility instead of being private or

b) also change `sched.rs` --> `schedulers.rs`

I am fine with a) , as we sort of have to trust everything in the Kernel crate anyway, so I am not that worried about visibility changes. If you prefer b), I would rather wait and do that in a followup PR, because changing the name of `sched.rs` would make this PR very difficult to review.

bradjc replied:
`sched.rs` should have been `kernel.rs` for a while now. Since after this PR it will definitely not do any scheduling, I think it is time to rename it to what it actually holds, the `Kernel` struct. And then it seems to make sense to have schedulers in a schedulers folder. We can make things pubcrate as needed.

hudson-ayers replied:
After this PR is merged, I will submit a PR that renames `sched.rs` to `kernel.rs` and renames the `sched/` folder to `schedulers/`, making visibility changes as needed.

bradjc commented on kernel/src/sched.rs:
```diff
@@ -49,6 +67,13 @@ pub struct Kernel {
 }
 
+#[derive(PartialEq, Eq)]
+pub(crate) enum StoppedExecutingReason {
```
Comments

ppannuto commented:
w.r.t. the proposed "universal" `kernel_loop`, as I understood it, one of the problems with that model as proposed are all of the calls to `service_pending_interrupts` and related interrupt methods. My understanding was that schedulers (may) need control of when interrupts are checked for and serviced?

@hudson-ayers Would it be possible to articulate the surface area of control knobs different schedulers need? I think that would help clarify the boundary.

hudson-ayers commented:
> w.r.t. the proposed "universal" `kernel_loop`, as I understood it, one of the problems with that model as proposed are all of the calls to `service_pending_interrupts` and related interrupt methods. My understanding was that schedulers (may) need control of when interrupts are checked for and serviced?

Yes, this was my concern with the universal `kernel_loop()` approach. Though Brad is right that if nothing enforces that you call `kernel_loop()` on the kernel, it is possible for different schedulers to reimplement their own kernel loop -- at the cost of inconsistent initialization in `main.rs`. 

> @hudson-ayers Would it be possible to articulate the surface area of control knobs different schedulers need? I think that would help clarify the boundary.

I think it is important that schedulers be able to choose when to service the bottom halves of interrupts (I am interested in delaying interrupt handling to help meet deadlines). I also have sort of vaguely imagined schedulers that toggle between different chip power modes based on apriori knowledge of process workloads, rather than relying on a single `sleep()` call, but that idea is less well defined and I don't think it is an absolute requirement for what we merge now.

@bradjc thoughts on this design?

```rust
pub trait Scheduler<C: Chip> {
    /// Return the next process to run and the timeslice length for that process.
    /// A timeslice of 0 means do not use the systick.
    fn next(&self) -> (Option<&'static dyn process::ProcessType>, u32, bool);

    /// Tell the scheduler to execute kernel work such as interrupt bottom halves and dynamic deferred calls
    /// Most schedulers will use this default implementation, but schedulers which at times wish to defer interrupt
    /// handling will reimplement it.
    fn execute_kernel_work(&self, chip: &C) {
        chip.service_pending_interrupts();
        DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
    }

    /// Ask the scheduler whether to take a break from executing userspace processes to handle kernel tasks
    /// Most schedulers will use this default implementations, but schedulers that wish to defer interrupt handling
    /// may reimplement it
    fn break_for_kernel_tasks(&self, kernel: Kernel, chip: &C) -> bool {
        chip.has_pending_interrupts()
        || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
        || kernel.processes_blocked()
    }

    /// Informs the scheduler of why the last process stopped executing.
    fn result(&self, result: StoppedExecutingReason, execution_time_us: u32);
}
```

```rust
pub fn kernel_loop<P: Platform, C: Chip, SC: Scheduler>(
        &self,
        platform: &P,
        chip: &C,
        ipc: Option<&ipc::IPC>,
        scheduler: &SC,
        _capability: &dyn capabilities::MainLoopCapability,
    ) -> ! {
        loop {
            unsafe {
                scheduler.execute_kernel_work();

                loop {
                    if scheduler.break_for_kernel_tasks() {
                        break;
                    }
                    let (process, timeslice_us, defer_interrupts) = scheduler.next();
                    process.map(|process| {
                        let reason =
                            self.do_process(platform, chip, process, ipc, timeslice_us, defer_interrupts);
                        scheduler.result(reason.0, reason.1);
                    });
                }

                chip.atomic(|| {
                    if !chip.has_pending_interrupts()
                        && !DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
                        && self.processes_blocked()
                    {
                        chip.sleep();
                    }
                });
            };
        }
    }
```

bradjc commented:
I really like the trait with default implementations, plus the ability to write your own kernel loop if you really want to do something radical.

I would ask that you copy what you have written here in this PR into the scheduler trait comments, so that future readers can get a sense of the thinking as to why the extra functions are stubbed out. It will also help us remember what we were thinking if a change is proposed in the future.

hudson-ayers commented:
Updated to use the trait with default implementations, rebased on master.

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -19,11 +27,49 @@ use crate::process::{self, Task};
+/// Trait which any scheduler must implement.
+pub trait Scheduler<C: Chip> {
+    /// Return the next process to run and the timeslice length for that process.
+    /// A timeslice of 0 means do preempt apps.
+    fn next(&self) -> (Option<AppId>, u32, bool);
```
I still think it makes more sense semantically for timeslice to be an `Option` rather than "0 means infinity"

hudson-ayers replied:
Done

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -19,11 +27,49 @@ use crate::process::{self, Task};
+    fn next(&self) -> (Option<AppId>, u32, bool);
+
+    /// Informs the scheduler of why the last process stopped executing, and how
+    /// long it executed for. Notabled execution_time_us will be innaccurate
```
```suggestion
    /// long it executed for. Notably execution_time_us will be inaccurate
```

alistair23 commented on boards/nordic/nrf52840_dongle/src/main.rs:
```diff
@@ -7,13 +7,15 @@
 use kernel::common::dynamic_deferred_call::{DynamicDeferredCall, DynamicDeferredCallClientState};
 use kernel::component::Component;
 #[allow(unused_imports)]
-use kernel::{capabilities, create_capability, debug, debug_gpio, debug_verbose, static_init};
```
Why move capabilities out here?

hudson-ayers replied:
good catch, fixed

hudson-ayers commented on boards/nordic/nrf52840_dongle/src/main.rs:
```diff
@@ -7,13 +7,15 @@
 use kernel::component::Component;
 #[allow(unused_imports)]
-use kernel::{capabilities, create_capability, debug, debug_gpio, debug_verbose, static_init};
+use kernel::{create_capability, debug, debug_gpio, debug_verbose, static_init};
```
```suggestion
use kernel::{capabilities, create_capability, debug, debug_gpio, debug_verbose, static_init};
```
good catch, that was an accident

hudson-ayers commented:
After some thought, I added one more function to the scheduler trait, with another default implementation. This allows for schedulers to specify the function that will determine when to break from the loop in `do_process()`. I think this approach is clearer than relying on a `break_for_interrupts` boolean, and allows schedulers to preempt apps at any point at which another process has an opportunity to become ready, which is important for priority schedulers which expect immediate preemption of lower priority apps, rather than waiting for a timeslice expiration.

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -19,11 +27,60 @@ use crate::process::{self, Task};
+        DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
+    }
+
+    /// Ask the scheduler whether to take a break from executing userspace
+    /// processes to handle kernel tasks. Most schedulers will use this
+    /// default implementations, but schedulers that wish to defer interrupt handling
+    /// may reimplement it.
```
```suggestion
    /// Ask the scheduler whether to take a break from executing userspace
    /// processes to handle kernel tasks. Most schedulers will use this
    /// default implementation, which always prioritizes kernel work, but
    /// schedulers that wish to defer interrupt handling may reimplement it.
```

bradjc commented on kernel/src/sched/round_robin.rs:
```diff
@@ -0,0 +1,84 @@
+//! Round Robin Scheduler for Tock
+//! This scheduler is specifically a Round Robin Scheduler with Interrupts.
```
```suggestion
//! This scheduler is specifically a Round Robin Scheduler with Interrupts.
//!
```

bradjc commented on kernel/src/sched/mlfq.rs:
```diff
@@ -0,0 +1,179 @@
+//! Multilevel feedback queue scheduler for Tock
```
```suggestion
//! Multilevel feedback queue scheduler for Tock
//!
```

bradjc commented on kernel/src/sched/cooperative.rs:
```diff
@@ -0,0 +1,63 @@
+//! Cooperative Scheduler for Tock
```
```suggestion
//! Cooperative Scheduler for Tock
//!
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -320,27 +402,57 @@ impl Kernel {
+    /// by looping within this function. This function will only return
+    /// control to the scheduler if a process yields with no callbacks pending,
+    /// exceeds its timeslice, or is interrupted.
+    /// Depending on the particular scheduler in use, this function can be configured
```
```suggestion
    ///
    /// Depending on the particular scheduler in use, this function can be configured
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -320,27 +402,57 @@ impl Kernel {
+    /// will use a dummy systick rather than systick of the chip in use. Schedulers can
+    /// pass a timeslice (in us) of their choice, though if the passed timeslice
+    /// is smalled than MIN_QUANTA_THRESHOLD_US the process will not execute, and
+    /// this function will return immediately.
```
```suggestion
    /// this function will return immediately.
    ///
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -19,11 +27,60 @@ use crate::process::{self, Task};
+    /// Tell the scheduler to execute kernel work such as interrupt bottom halves
+    /// and dynamic deferred calls. Most schedulers will use this default
+    /// implementation, but schedulers which at times wish to defer interrupt
+    /// handling will reimplement it.
```
```suggestion
    /// handling will reimplement it.
    ///
```

bradjc commented on kernel/src/process.rs:
```diff
@@ -262,6 +262,9 @@ pub trait ProcessType {
     /// execute any new tasks.
     fn enqueue_task(&self, task: Task) -> bool;
 
+    /// Returns whether this process is ready to execute
```
```suggestion
    /// Returns whether this process is ready to execute.
```

bradjc commented on boards/acd52832/src/main.rs:
```diff
@@ -4,6 +4,7 @@
 // Disable this attribute when documenting, as a workaround for
 // https://github.com/rust-lang/rust/issues/62184.
 #![cfg_attr(not(doc), no_main)]
+#![feature(const_in_array_repeat_expressions)]
```
?

hudson-ayers replied:
explained here: https://github.com/tock/tock/pull/1767#issuecomment-645646829

I can add a comment saying as much in every board file

hudson-ayers commented on kernel/src/sched.rs:
```diff
@@ -19,11 +27,60 @@ use crate::process::{self, Task};
+    /// Informs the scheduler of why the last process stopped executing, and how
+    /// long it executed for. Notably, execution_time_us will be 0
+    /// if the the scheduler requested this process be run cooperatively.
+    fn result(&self, result: StoppedExecutingReason, execution_time_us: u32);
```
you're right, fixed.

hudson-ayers commented on kernel/src/platform/systick.rs:
```diff
@@ -53,6 +65,10 @@ impl SysTick for () {
     }
 
+    fn get_value(&self) -> u32 {
+        10000 // this is a hack, basically just a value larger than the MIN_QUANTA_THRESHOLD is expected to be
```
done

alistair23 commented on boards/components/src/sched/mlfq.rs:
````diff
@@ -0,0 +1,97 @@
+//! ```
+
+// Author: Hudson Ayers <hayers@stanford.edu>
+// Last modified: 03/31/2020
````
This is already wrong, it's probably worth removing.

hudson-ayers commented:
Update: I think I have addressed all reviews here.

I am waiting on #1985 being merged, as it seems very close, then I will resolve the merge conflicts that will create with this PR and test it again.

hudson-ayers commented:
I have squashed and rebased this PR following #1985 / #1887 . I have tested it as follows on Imix:

1. Using Round robin, run blink + hello_loop + whileone -- works. Timeslice expirations only detected for whileone (checked via process console).

2. Using cooperative scheduler, run blink + hello_loop + whileone -- hangs as expected

3. Using MLFQ scheduler, run blink + hello_loop + whileone -- works. Also, I see a timeslice expiration pattern similar to what I would expect given cumulative timeslice tracking + redemptions.

4. Using priority scheduler, run blink + hello_loop + whileone -- works if whileone is the lowest priority app, hangs as expected if whileone is the highest priority app

This should be ready for review!

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -653,6 +771,14 @@ impl Kernel {
             }
         }
+        let time_executed_us = timeslice_us.map_or(None, |timeslice| {
+            // min is to protect from wrapping
+            Some(u32::min(
+                timeslice - scheduler_timer.get_remaining_us(),
+                timeslice,
```
I had trouble reasoning about when this could possibly wrap, however one scenario I came up with was a coarse-grained timer, in which case the `remaining_us` might be greater than the timeslice if the process executed for a very _short_ window of time; but this correction will determine that the _whole_ timeslice was used rather than almost none of it.

I'm not sure if that's the wrapping that's being protected against? If it is, I don't think this is the right logic. In either case, at least a bit more comment explanation here would help I think.

hudson-ayers replied:
I was thinking of the `VirtualSchedulerTimer` when I wrote this -- I am concerned about the case where the timer wraps around after we return from userspace, but the process has yielded, so we return without checking `scheduler_timer.has_expired()` again. For the virtual timer, the time returned will be very large in this case, so I wanted to coerce any very large values to 0...which this does not do. good catch! I changed it to check `has_expired()` after calling `get_remaining_us()` instead, that is much less brittle.

I also updated the documentation of `get_remaining_us()` to make it more clear why this is needed.

bradjc commented on doc/Startup.md:
```diff
@@ -179,5 +179,8 @@ invalid TBF header in flash.
+Tock provides a `Scheduler` trait that serves as an abstraction to allow for
+plugging in different scheduling algorithms. Schedulers should be initialized
+at the end of the reset handler.
+The final thing that the reset handler must do is call `scheduler.kernel_loop()`.
```
```suggestion
The final thing that the reset handler must do is call `kernel.kernel_loop()`.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -1,4 +1,12 @@
-//! Tock core scheduler.
+//! Defines the central Kernel struct and a trait that
+//! different scheduler implementations must implement. Also defines several
+//! utility functions to reduce repeated code between different scheduler
+//! implementations.
```
```suggestion
//! Tock's central kernel logic and scheduler trait.
//!
//! Also defines several
//! utility functions to reduce repeated code between different scheduler
//! implementations.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +28,60 @@ use crate::process::{self, Task};
+
+/// Trait which any scheduler must implement.
+pub trait Scheduler<C: Chip> {
+    /// Return the next process to run, and the timeslice length for that process.
+    /// Passing None instead of an AppId will lead to nothing being run.
+    /// Passing None instead of a timeslice will cause the process to be run cooperatively
```
```suggestion
    /// Return the next process to run, and the timeslice length for that process.
    ///
    /// The first argument is an optional `AppId` of the process to run next.
    /// Returning `None` instead of `Some(AppId)` means nothing will run. The second
    /// argument is an optional number of microseconds to use as the length of the process's
    /// timeslice. If an `AppId` is returned in the first argument, then returning `None`
    /// instead of a timeslice will cause the process
    /// to be run cooperatively (i.e. without preemption). Otherwise the process will run
    /// with a timeslice set to the specified length.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +28,60 @@ use crate::process::{self, Task};
+    /// Passing None instead of a timeslice will cause the process to be run cooperatively
+    fn next(&self) -> (Option<AppId>, Option<u32>);
+
+    /// Informs the scheduler of why the last process stopped executing, and how
+    /// long it executed for. Notably, execution_time_us will be `None`
+    /// if the the scheduler requested this process be run cooperatively.
```
```suggestion
    /// Inform the scheduler of why the last process stopped executing, and how
    /// long it executed for. Notably, `execution_time_us` will be `None`
    /// if the the scheduler requested this process be run cooperatively.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +28,60 @@ use crate::process::{self, Task};
+    /// implementation, but schedulers which at times wish to defer interrupt
+    /// handling will reimplement it.
+    ///
+    /// The reason for stubbing this out to allow schedulers to reimplement it
+    /// is that I (Hudson Ayers) believe that deferring bottom half
+    /// interrupt handling, or only servicing certain interrupts,
+    /// can be useful in helping userspace apps to meet
+    /// deadlines if a scheduler is given sufficient information about app
+    /// workloads. This stub could also potentially be useful for a power aware
+    /// scheduler which toggles between chip power modes based on preexisting knowledge
+    /// of process workloads.
```
```suggestion
    /// Providing this interface allows schedulers to fully manage how
    /// the main kernel loop executes. For example, a more advanced
    /// scheduler that attempts to help processes meet their deadlines may
    /// need to defer bottom half interrupt handling or to selectively service
    /// certain interrupts. Or, a power aware scheduler may want to selectively
    /// choose what work to complete at any time to meet power requirements.
    ///
    /// Custom implementations of this function must be very careful, however,
    /// as this function is called in the core kernel loop.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +28,60 @@ use crate::process::{self, Task};
+        DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
+    }
+
+    /// Ask the scheduler whether to take a break from executing userspace
+    /// processes to handle kernel tasks. Most schedulers will use this
+    /// default implementation, which always prioritizes kernel work, but
+    /// schedulers that wish to defer interrupt handling may reimplement it.
+    unsafe fn break_for_kernel_tasks(&self, kernel: &Kernel, chip: &C) -> bool {
+        chip.has_pending_interrupts()
+            || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
+            || kernel.processes_blocked()
+    }
```
The interaction between `next()` and `break_for_kernel_tasks()` is a little surprising. Namely, I was expecting that if a scheduler returns `None` from `next()`, that seems like an indication for the chip to sleep. Instead, by default, the kernel will just continuously call `next()`.

Also, `break_for_kernel_tasks()` is actually `break_for_kernel_tasks_or_to_sleep()`.

What about removing `break_for_kernel_tasks()`, and adding the break to `next().map_or_else(|| break; ...`?

hudson-ayers replied:
The advantage of keeping `break_for_kernel_tasks()` is that all of the current schedulers use the default implementation, and it makes it very clear that the checks included in the default implementation should typically be used. 

I seem to recall you not liking when that code was re-implemented for each scheduler, which would be the case if it was moved into `next()`. But I agree the current interaction is confusing -- I am happy to take the approach of moving it into `next()` if you think that will be maintainable

bradjc replied:
Ok I see, and yeah I didn't think about how that would force that duplicated logic into all of the `next()` calls. There are three clear decisions for the scheduler to make:

1. What process to run next, or none.
2. Whether to stop looking at processes and instead do kernel work.
3. Whether to continue to allow the current process to run.

And one more decision the kernel has to make:

4. Whether to go to sleep.

The three clear decisions each should be a function, like you have. I think what we need to establish is how does a scheduler decide if the chip should go to sleep, and how much control does it have, versus the kernel struct making some decisions to in theory help robustness.

Items (1) and (2) are related.

```
next()    | break_for_work()    | effect
-------------------------------------------
x         | true                | processes are not executed
some      | false               | that process is executed
none      | false               | ?
```

I believe right now the "?" is sit in a busy loop. I think instead that should be "try to sleep".

bradjc replied:
In my proposed setup, if a scheduler returns `true` from `break_for_kernel_work()`, then we have to promise that `next()` will be called again before trying to sleep. That might be tricky to implement.

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +28,60 @@ use crate::process::{self, Task};
+    /// However, schedulers which wish to defer interrupt handling may change this, or
+    /// priority schedulers which wish to check if the execution of the current process
+    /// has caused a higher priority process to become ready (such as in the case of IPC)
+    unsafe fn leave_do_process(&self, chip: &C) -> bool {
```
```suggestion
    unsafe fn continue_process(&self, process: AppId, chip: &C) -> bool {
```

Name change to try to better reflect what it is asking of the scheduler. Then, do you think it makes sense to tell the kernel what process it is asking about?

hudson-ayers replied:
Agreed, changed.

bradjc commented on kernel/src/sched.rs:
```diff
@@ -50,11 +108,28 @@ pub struct Kernel {
+/// Enum used to inform scheduler why a process stopped
+/// executing (aka why `do_process()` returned).
+#[derive(PartialEq, Eq)]
+pub enum StoppedExecutingReason {
+    /// The process returned because it is no longer ready to run
+    NoWorkLeft,
+
+    /// The process returned because its timeslice expired
+    TimesliceExpired,
+
+    /// The process returned because it was preempted by kernel
+    /// work that became ready (most likely because an interrupt fired
+    /// and the kernel thread needs to execute the bottom half of the
+    /// interrupt)
+    KernelPreemption,
+}
```
What about it crashed or exited?

hudson-ayers replied:
Those are currently grouped in as part of `NoWorkLeft`, as none of the schedulers I have implemented do anything different for crashed/exited than they do for yielded. But it makes sense that some schedulers might want to handle those cases differently (I could imagine optimization of scheduler internal data structures by removing nodes that won't be scheduled, or a scheduler punishing crashed processes even after they restart), so I will add them in.

hudson-ayers replied:
Though, I am not sure adding "Exited" makes sense until Tock 2.0 when we have an Exited syscall

hudson-ayers replied:
Added `StoppedFaulted` and `Stopped`. I specified `StoppedFaulted` bc restarting faults will not cause do_process to exit

bradjc commented on kernel/src/sched.rs:
```diff
@@ -373,28 +473,62 @@ impl Kernel {
+                break;
+            }
+
+            if scheduler.continue_process(process.appid(), chip) {
```
```suggestion
            if !scheduler.continue_process(process.appid(), chip) {
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +29,69 @@ use crate::process::{self, Task};
+    /// processes to handle kernel tasks. Most schedulers will use this
+    /// default implementation, which always prioritizes kernel work, but
+    /// schedulers that wish to defer interrupt handling may reimplement it.
+    unsafe fn break_for_kernel_tasks(&self, kernel: &Kernel, chip: &C) -> bool {
```
```suggestion
    unsafe fn break_for_kernel_work(&self, kernel: &Kernel, chip: &C) -> bool {
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -373,28 +473,62 @@ impl Kernel {
         }
```
With this implementation, the scheduler cannot force the chip to go to sleep. I think at least the `&& self.processes_blocked()` check needs to be removed, otherwise a power scheduler (say) could not decide to sleep while a process is runnable if it is trying to save energy.

For interrupts (and software interrupts), I don't think the decision is as clear. If a scheduler ignores pending interrupts, will the chip just immediately wake up (or not sleep at all) on some (all?) platforms? Also, if there are no other pending interrupts, a scheduler could cause the system to stop as it would never wake up again.

I think making a conscious decision to say that the scheduler (at least right now) cannot ignore pending interrupts to go to sleep is reasonable, but we should document that.

hudson-ayers replied:
I have pushed an update that resolves this by removing the nested loop so that there is only a single main loop that matches whether to perform kernel work or try to run a process / sleep. I actually think that a design like this might be cleaner than what I just pushed:
```
/// Enum representing the actions the scheduler can request in each call to scheduler.next()
#[derive(Copy, Clone)]
pub(crate) enum SchedulingDecision {
    /// Tell the kernel loop to call execute_kernel_tasks()
    KernelWork,

    /// Tell the kernel to run the specified process with the passed timeslice.
    /// If `None` is passed as a timeslice, the process will be run cooperatively.
    RunProcess((AppId, Option<u32>)),

    /// Tell the kernel to go to sleep. Notably, if the scheduler asks the kernel
    /// to sleep when kernel tasks are ready, the kernel will not sleep, and will
    /// instead restart the main loop and call `next()` again.
    Sleep,
}

impl Kernel {
    unsafe fn kernel_work_ready(&self, chip: &C) -> bool {
        chip.has_pending_interrupts()
            || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
    }

    fn kernel_loop(...) {
        loop {
            chip.watchdog().tickle();
            unsafe {
                scheduler.execute_kernel_work(chip);
                match scheduler.next(self, kernel, chip) {
                    KernelWork => scheduler.execute_kernel_work(chip),
                    RunProcess((appid, timeslice)) => {
                        self.process_map_or((), appid, |process| {
                            let (reason, time_executed) = self.do_process(
                                platform,
                                chip,
                                scheduler,
                                process,
                                ipc,
                                timeslice_us,
                            );
                            scheduler.result(reason, time_executed);
                        });
                    }
                    Sleep => {
                        chip.atomic(|| {
                            // cannot sleep if interrupts are pending, as on most platforms
                            // unhandled interrupts will wake the device.
                            if !chip.has_pending_interrupts()
                                && !DynamicDeferredCall::global_instance_calls_pending()
                                    .unwrap_or(false)
                            {
                                chip.watchdog().suspend();
                                chip.sleep();
                                chip.watchdog().resume();
                            }
                        });
                    }
                }
            }
        }
    }
}
```

But then we still rely on each scheduler implementation calling `kernel.kernel_work_read()` within `next()`, rather than the cleaner trait function with a default implementation.  Thoughts?

ppannuto replied:
What's the purpose of the `scheduler.execute_kernel_work(chip);` _before_ the call to `next`? Couldn't that be the call to `kernel_work_ready`?

hudson-ayers replied:
It shouldn't be there, I just forgot to delete that line. Edited!

bradjc replied:
I think I prefer the current approach in the PR, but with next() returning an enum like you have in the comment (but with only run and sleep as options). That way the structure of doing kernel work first is preserved, and a simpler scheduler doesn't have to worry about getting that right.

It also keeps the narrow scope of `next()`: the kernel wants to run a process, the scheduler has to say which one or none. What happens if a scheduler _really_ wants to prioritize kernel work? Well, I'm not sure Tock is a great fit for that. At some point, we can't just continuously check if an interrupt has triggered. Once a scheduler returns `RunProcess`, the kernel will call `continue_process()` before actually running anything, so the scheduler could stop there and resume kernel work. If an interrupt fires after then but before the actual context switch, the interrupt won't be handled until after the process runs. I think that is reasonable.

hudson-ayers replied:
Yeah, and the scheduler can redefine `execute_kernel_work()` to loop if required, so I think this design is fine. I pushed an update that uses an enum with only run and sleep options.

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +29,69 @@ use crate::process::{self, Task};
+    unsafe fn break_for_kernel_tasks(&self, kernel: &Kernel, chip: &C) -> bool {
+        chip.has_pending_interrupts()
+            || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
+            || kernel.processes_blocked()
```
```suggestion
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -332,31 +424,39 @@ impl Kernel {
-    ) {
-        chip.watchdog().setup();
+    ) -> ! {
         loop {
```
I am having a hard time reasoning about if this loop can be constructed according to my other comments.

hudson-ayers replied:
I think the solution is to not have a nested loop. The nested loop isn't really needed and makes reasoning about program flow a lot more confusing.

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +29,81 @@ use crate::process::{self, Task};
+            || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
+    }
+
+    /// Ask the scheduler whether to return from the loop in `do_process()`. Most
+    /// schedulers will use this default implementation, which causes the do_process()
+    /// loop to return if there are interrupts or deferred calls that need to be serviced.
+    /// However, schedulers which wish to defer interrupt handling may change this, or
+    /// priority schedulers which wish to check if the execution of the current process
+    /// has caused a higher priority process to become ready (such as in the case of IPC)
+    /// The `AppId` of the process that is executing is passed in the event that future
+    /// schedulers need this info when reimplementing this.
```
```suggestion
    /// Ask the scheduler whether to continue trying to execute a process.
    ///
    /// Once a process is scheduled the kernel will try to execute it until it has no more
    /// work to do or exhausts its timeslice. The kernel will call this function before
    /// every loop to check with the scheduler if it wants to continue trying to execute
    /// this process.
    ///
    /// Most
    /// schedulers will use this default implementation, which causes the `do_process()`
    /// loop to return if there are interrupts or deferred calls that need to be serviced.
    /// However, schedulers which wish to defer interrupt handling may change this, or
    /// priority schedulers which wish to check if the execution of the current process
    /// has caused a higher priority process to become ready (such as in the case of IPC).
    ///
    /// `id` is the identifier of the currently active process.
```

ppannuto replied:
```suggestion
    /// schedulers need this info when reimplementing this.
    ///
    /// If this returns `false`, then `do_process` will exit with a `KernelPreemption`.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -20,10 +29,81 @@ use crate::process::{self, Task};
+    }
+}
+
+/// Enum representing the actions the scheduler can request in each call to scheduler.next()
```
```suggestion
/// Enum representing the actions the scheduler can request in each call to `scheduler.next()`.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -332,69 +436,117 @@ impl Kernel {
+                            }
+                            SchedulingDecision::Sleep => {
+                                chip.atomic(|| {
+                                    // cannot sleep if interrupts are pending, as on most platforms
+                                    // unhandled interrupts will wake the device.
```
```suggestion
                                    // Cannot sleep if interrupts are pending, as on most platforms
                                    // unhandled interrupts will wake the device. Also, if the only pending
                                    // interrupt occurred after the scheduler decided to put the chip
                                    // to sleep, but before this atomic section starts, the interrupt will
                                    // not be serviced and the chip will never wake from sleep.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -50,11 +130,35 @@ pub struct Kernel {
+    /// The process returned because its timeslice expired
+    TimesliceExpired,
+
+    /// The process returned because it was preempted by kernel
+    /// work that became ready (most likely because an interrupt fired
+    /// and the kernel thread needs to execute the bottom half of the
+    /// interrupt)
```
```suggestion
    /// The process returned because it was preempted by the kernel.
    /// This can mean that kernel work became ready (most likely because an interrupt fired
    /// and the kernel thread needs to execute the bottom half of the
    /// interrupt), or because the scheduler no longer wants to execute that
    /// process.
```

bradjc commented on kernel/src/sched.rs:
```diff
@@ -332,69 +436,117 @@ impl Kernel {
+                                        timeslice_us,
+                                    );
+                                    scheduler.result(reason, time_executed);
+                                });
```
I think this is just a rust thing we have to deal with. A buggy scheduler that returns non existent `AppId`s doesn't seem like a case worth checking for.

hudson-ayers replied:
Yeah generally agreed this is another case for #1834 . We could add a `process_map_or_else()` and then panic in the missing case, I guess?

bradjc replied:
That would protect against two cases:
1. The process moves in the array, or is removed, or otherwise stops existing or
2. The scheduler just made up an AppId.

The first is why process_map_or exists, but how could something happen to a process between the scheduler choosing from the list and the next line of code? And the second option could happen, I guess, but why would we add a panic!() output? It should be pretty obvious that the scheduler is calling `AppId::new()` and the scheduler doesn't work.

hudson-ayers replied:
yeah, lets just leave it. The only reason to change it is that the `map_or((), ...)` is a bit of a code smell, but there is an actual justification here.

bradjc commented on kernel/src/sched.rs:
```diff
@@ -332,69 +436,117 @@ impl Kernel {
+                            }
+                            SchedulingDecision::Sleep => {
+                                chip.atomic(|| {
+                                    // cannot sleep if interrupts are pending, as on most platforms
+                                    // unhandled interrupts will wake the device.
+                                    if !chip.has_pending_interrupts()
+                                        && !DynamicDeferredCall::global_instance_calls_pending()
+                                            .unwrap_or(false)
```
Isn't the other case that the chip never wakes up again?

ppannuto replied:
I think the other case is perhaps that the chip doesn't wake until a _different_ interrupt triggers (e.g. the scheduler chooses to ignore a peripheral, go to sleep, but wants to wake for a timer)

hudson-ayers replied:
I don't think we could just remove the check, because of the scenario where an interrupt arrives while `scheduler.next()` is executing and returning `Sleep`. We could add *another* scheduler trait function with a default impl there, if we want to support sleeping despite outstanding (pending but disabled) interrupts. I don't feel super strongly either way

bradjc replied:
I propose that if a use case comes around that demands that level of control, we could revisit this, but I think it is good to have some protective defaults at this low layer.

hudson-ayers commented on kernel/src/sched.rs:
```diff
@@ -20,10 +29,81 @@ use crate::process::{self, Task};
+    /// Tell the kernel to go to sleep. Notably, if the scheduler asks the kernel
+    /// to sleep when kernel tasks are ready, the kernel will not sleep, and will
+    /// instead restart the main loop and call `next()` again.
+    Sleep,
```
Maybe `TrySleep`? I like that a little better, just because `RunNoProcesses` doesn't really make it clear that sleeping is the likely outcome.

hudson-ayers replied:
Changed to `TrySleep`

hudson-ayers commented on kernel/src/sched.rs:
```diff
@@ -20,10 +29,81 @@ use crate::process::{self, Task};
+    /// processes to handle kernel tasks. Most schedulers will use this
+    /// default implementation, which always prioritizes kernel work, but
+    /// schedulers that wish to defer interrupt handling may reimplement it.
+    unsafe fn kernel_work_ready(&self, chip: &C) -> bool {
```
changed

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -1,4 +1,13 @@
-//! Tock core scheduler.
+//! Tock's central kernel logic and scheduler trait.
+//!
+//! Also defines several
+//! utility functions to reduce repeated code between different scheduler
+//! implementations.
```
```suggestion
//! Also defines several utility functions to reduce repeated 
//! code between different scheduler implementations.
```

ppannuto commented on kernel/src/sched.rs:
```diff
@@ -332,69 +446,122 @@ impl Kernel {
-                        chip.sleep();
-                        chip.watchdog().resume();
+                    false => {
+                        // No kernel work ready, so ask scheduler for a process
```
```suggestion
                        // No kernel work ready, so ask scheduler for a process.
```

bradjc commented:
bors r+
